src/library_exceptions.js

```diff
@@ -7,92 +7,196 @@
+    }
+
+    this.set_caught = function (caught) {
+      caught = caught ? 1 : 0;
+      {{{ makeSetValue('this.ptr', '___ExceptionInfoAttrs.CAUGHT_OFFSET', 'caught', 'i8') }}};
+    }
+
+    this.get_caught = function () {
+      return {{{ makeGetValue('this.ptr', '___ExceptionInfoAttrs.CAUGHT_OFFSET', 'i8') }}} != 0;
+    }
+
+    this.set_rethrown = function (rethrown) {
+      rethrown = rethrown ? 1 : 0;
+      {{{ makeSetValue('this.ptr', '___ExceptionInfoAttrs.RETHROWN_OFFSET', 'rethrown', 'i8') }}};
+    }
+
+    this.get_rethrown = function () {
+      return {{{ makeGetValue('this.ptr', '___ExceptionInfoAttrs.RETHROWN_OFFSET', 'i8') }}} != 0;
+    }
+
+    this.init = function(type, destructor) {
+      this.set_type(type);
+      this.set_destructor(destructor);
+      this.set_refcount(0);
+      this.set_caught(false);
+      this.set_rethrown(false);
+    }
+
+    this.add_ref = function() {
+#if USE_PTHREADS
+      Atomics.add(HEAP32, (this.ptr + ___ExceptionInfoAttrs.REFCOUNT_OFFSET) >> 2, 1);
+#else
+      var value = {{{ makeGetValue('this.ptr', '___ExceptionInfoAttrs.REFCOUNT_OFFSET', 'i32') }}};
+      {{{ makeSetValue('this.ptr', '___ExceptionInfoAttrs.REFCOUNT_OFFSET', 'value + 1', 'i32') }}};
 #endif
-          return ptr;
-        }
+    }
+
+    // Returns true if last reference released.
+    this.release_ref = function() {
+#if USE_PTHREADS
+      var prev = Atomics.sub(HEAP32, (this.ptr + ___ExceptionInfoAttrs.REFCOUNT_OFFSET) >> 2, 1);
+#else
+      var prev = {{{ makeGetValue('this.ptr', '___ExceptionInfoAttrs.REFCOUNT_OFFSET', 'i32') }}};
+      {{{ makeSetValue('this.ptr', '___ExceptionInfoAttrs.REFCOUNT_OFFSET', 'prev - 1', 'i32') }}};
+#endif
+      if (prev === 0) {
+        err('Exception reference counter underflow');
       }
+      return prev === 1;
     }
-#if EXCEPTION_DEBUG
-    err('no de-adjustment for unknown exception ptr ' + adjusted);
+  },
+
+  CatchInfo__deps: ['ExceptionInfo'],
+  // Returned from __cxa_find_matching_catch, stores base and adjusted exception pointer.
+  CatchInfo: function(ptr) {
+
+    this.free = function() {
+      _free(this.ptr);
+      this.ptr = 0;
+    }
+
+    this.set_base_ptr = function(basePtr) {
+      {{{ makeSetValue('this.ptr', '0', 'basePtr', '*') }}};
+    }
+
+    this.get_base_ptr = function() {
+      return {{{ makeGetValue('this.ptr', '0', '*') }}};
+    }
+
+    this.set_adjusted_ptr = function(adjustedPtr) {
+      var ptrSize = {{{ Runtime.POINTER_SIZE }}};
+      {{{ makeSetValue('this.ptr', 'ptrSize', 'adjustedPtr', '*') }}};
+    }
+
+    this.get_adjusted_ptr = function() {
+      var ptrSize = {{{ Runtime.POINTER_SIZE }}};
+      return {{{ makeGetValue('this.ptr', 'ptrSize', '*') }}};
+    }
+
+    this.get_exception_ptr = function() {
+#if WASM_BACKEND || DISABLE_EXCEPTION_CATCHING != 1 // work around a fastcomp bug
+      var isPointer = {{{ exportedAsmFunc('___cxa_is_pointer_type') }}}
+        (this.get_exception_info().get_type());
```

Comment:
please put the `(` on the previous line. (I'm actually unsure what JS ASI does on something like this, but it worries me...)